src/allmydata/storage_client.py

```diff
@@ -461,6 +485,74 @@ def get_nickname(self):
         return "?"
 
 
+def parse_grid_manager_data(gm_data):
+    """
+    :param gm_data: some data that might be JSON that might be a valid
+       Grid Manager Certificate
+
+    :returns: json data of a valid Grid Manager certificate, or an
+        exception if the data is not valid.
+    """
+
+    required_keys = allowed_keys = [
+        'certificate',
+        'signature',
+    ]
+
+    js = json.loads(gm_data)
+    for k in js.keys():
+        if k not in allowed_keys:
+            raise ValueError(
+                "Grid Manager certificate JSON may not contain '{}'".format(
+                    k,
+                )
+            )
+    for k in required_keys:
+        if k not in js:
+            raise ValueError(
+                "Grid Manager certificate JSON must contain '{}'".format(
+                    k,
+                )
+            )
+    return js
+
+
+def validate_grid_manager_certificate(gm_key, alleged_cert, now_fn=None):
+    """
+    :param gm_key: a VerifyingKey instance, a Grid Manager's public
+        key.
+
+    :param alleged_cert: dict with "certificate" and "signature" keys, where
+        "certificate" contains a JSON-serialized certificate for a Storage
+        Server (comes from a Grid Manager).
+
+    :param now_fn: a zero-argument callable that returns a UTC
+        timestamp (will use datetime.utcnow by default)
+
+    :return: False if the signature is invalid or the certificate is
+        expired.
+    """
+    if now_fn is None:
+        now_fn = datetime.utcnow
+
+    try:
+        ed25519.verify_signature(
+            gm_key,
+            base32.a2b(alleged_cert['signature'].encode('ascii')),
+            alleged_cert['certificate'].encode('ascii'),
+        )
+    except ed25519.BadSignature:
+        return False
```

Comment:
There also seems to be many cases where revealing different error-codes gives attackers of the crypto more information, and is thus bad. Not sure I see how that might actually apply here, though -- perhaps another one for "ask our cryptographers"?

(That aside, how should this be expressed to the user? A log message here? Or I guess returning something besides true/false and a log message a little  higher up?)

Reply:
Just considering the last question for the moment, some kind of status system exists for storage servers.  eg, the web ui will show you all the storage servers a node knows about and some minimal information about them, like whether the node is currently connected to them or not.  "this node had a bogus signature" could be revealed by that system, perhaps.  I don't know if that stuff is just glued right into the web ui (not great) or if it's part of a more general system where it could be revealed by some CLI command or HTTP API (that would be nice).

Reply:
Looks like it's not really "a system" -- the Web stuff reaches in and asks (so, could add e.g. "upload_permitted?" column or something but it'd be a web-thing).

Reply:
https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3506